### ISSUE TITLE
Simultaneous translation of multiple items

### ISSUE THREAD

**Issue by sdlmw** [ModernMT/MMT#381]
Hi  MMT term，

  Is there a scheme to send a JSON containing multiple text，Returns the translation of each piece of text.  Instead of a single translation every time by GET way.

Thanks

**Comment by davidecaroselli** [comment]
Hi @sdlmw ,

ModernMT is built with real-time translation in mind, so its process is not suitable for batch translation. For this reason there is no such API, you will have to call a GET for every line you want to translate.

Best

**Comment by davidecaroselli** [comment]
By the way, if your goal is to run multiple translations at the same time, this is absolutely possibile: your client should implement a "pool" of translation requests made in parallel: if your engine has multiple GPUs you will be able to actually translate multiple sentences at the same time.

This is, for example, what it is done when you call `./mmt translate --batch` or even `./mmt evaluate`

**Comment by sdlmw** [comment]
@davidecaroselli  all right. Thank you for your serious answer.

**Comment by sdlmw** [comment]
Hi @davidecaroselli ，Whether you can add this feature to your work schedule？

**Comment by davidecaroselli** [comment]
Hi @sdlmw 

because MMT is at this time not able to leverage batch translation, the time needed for process a batch will be the same as requesting multiple GET (plus negligible network time i guess).
For this reason we don't see any advantages into introducing a "multiple translation" API, or am I missing something? What would be your use case?

**Comment by nonsequitir** [comment]
@davidecaroselli   There is an advantage in having large-scale translations handled in batch mode.  I totally appreciate your real-time comments and the REST interface is FAST today, but in a large service provider, one might be translating large documents through a translation management system with ModernMT in an automated workflow that could scale to 100+ users and many, many projects.  I just ran tests for a 500K word project in 10 languages.  

 Ideally, we'd parse these documents and aggregated their segments into chunks of 50Kb or so for batch translation to minimize network chatter and allow for future support in request throttling/queuing etc.   

Perhaps separating the RESTful server from the core code would allow for someone to contribute to the project with a solution for such an interface?  What's there already is terrific, but I'm thinking we could take over the world :)

**Comment by davidecaroselli** [comment]
Hi @nonsequitir 

I totally got your point, but let me comment on this.

On a **low level** perspective, each translation is different to ModernMT because of its adaptation feature: the Neural model is changed every time for every request to enhance the performance for every single translation. This of course has the drawback that, with the current implementation, it is impossible to translate a batch of sentences, because every single sentence needs its "version" of the neural model. This will probably change in the future, as we will improve the adaptation algorithm in this direction and at then, we will definitely add the batch-translation API.

For this reason, as you said, the only advantages would be network time for every single GET request.

Now the scenario you suggested (hundreds of users, and thousands of projects) is the most interesting one, and let me say: ModernMT already handles it to the best :) But you said it right: the OpenSource ModernMT on its own is not enough but, most important, it is the fundamental building block. Our Enterprise Edition service relies on the OpenSource ModernMT version, and adds a layer to handle user privacy/authentication/document translation and so on. 

**In conclusion**: my opinion is that the best thing to do is to add this API when the core itself will be able to leverage this feature, on the contrary we will just end up adding an API that won't help that much, and in the end, probably even in a real-case scenario you already need to implement a much more complex layer for user and document management that really depends on what are your needs.

**Comment by nonsequitir** [comment]
@davidecaroselli Totally agree, and I quite forgot about the Enterprise Edition :)

**Comment by sdlmw** [comment]
Hi @davidecaroselli 
I studied the translate module, but  didn't learn anything new. Now，i need to add a shell that can be multipled translations in http://localhost:8045/translate file, which file should I modify? 

Expect your reply，
Thank you

**Comment by davidecaroselli** [comment]
Hi @sdlmw 

if you are looking for a shell script that batch translate a file, it is already there:
```bash
./mmt translate --batch < your/source/file.en > output.it
```
This will translate a file line by line in a batch fashion. If you are looking for a XLIFF batch translator, this is the command:
```bash
./mmt translate --xliff < your/source/file.xliff > output.xliff
```
If you are looking for a custom script, a good starting point is the python file [translation.py](https://github.com/ModernMT/MMT/blob/master/cli/translation.py) where the class [BatchTranslator](https://github.com/ModernMT/MMT/blob/master/cli/translation.py#L93) should be exactly what you are looking for.

Davide

**Comment by sdlmw** [comment]
Hi @davidecaroselli  
I am very sorry , According to your 3rd, I started to try.  First I added the file log output to each function. The things that make me confused. Does not generate any file log that I have defined. I deleted the Translation.pyc file, and retry. The problem is still. Next, I looked up the URL based on。I found RESTServer$Router.class. But this is a Java-compiled file. Interrupted again
So，Could you let me know the meaning of each parameter inside BatchTanslator class?

Thanks

**Comment by sdlmw** [comment]
HI @davidecaroselli 

./mmt translate --context "programming language tutorial" "hello world" by the command.  I found the right place. But I need to change the http://localhost:8045/translate way, Could you explain the relevant documents to me?

thanks

**Comment by davidecaroselli** [comment]
Hi @sdlmw 

sure! So the Java source file corresponding to that API is this one: [Translate.java](https://github.com/ModernMT/MMT/blob/master/src/api-rest/src/main/java/eu/modernmt/rest/actions/translation/Translate.java)
After some parameters parsing, it forward the call to the Facade interface of ModernMT, and the method invoked is this one: [TranslationFacade.get()](https://github.com/ModernMT/MMT/blob/master/src/core/src/main/java/eu/modernmt/facade/TranslationFacade.java#L100)

Hope this helps!

**Comment by davidecaroselli** [comment]
Hi @sdlmw 

I'm closing this issue for now but if you have any other question please don't hesitate to re-open it!

**Comment by bittlingmayer** [comment]
@sdlmw 

For our purposes - hammering the Enterprise Edition API - "batching" would be ideal.

Right now, to get the best results, we would need to include the context.  So what comes to our server as a batch of 10 originals gets send to ModernMT `/translate` 1 at a time, and the context has to get rolled into each.

So the total size of the requests ends up 10x larger than necessary, when really not that much unique information is being sent.  It would make more sense for the "context expansion" to just be done by ModernMT.

Especially because if the context is not 10 lines but a 1000-line document - which is going to happen more and more in the future - then it's 1000x.

Maybe it should be just an option, and I would be fine if the default were to consider the `q`s in a batch independently.

I understand your point that you can't necessarily create context as optimally as the client applications can, but it's probably equal quality and better performance (and cleaner code) for the 90%, and those with special cases would be free to do as they do now.

**Comment by bittlingmayer** [comment]
Update:

This is now supported.

<img width="1123" alt="Screenshot 2021-03-22 at 21 16 20" src="https://user-images.githubusercontent.com/11457984/112030662-ed04e700-8b53-11eb-9680-07cfec6d0d92.png">

**Comment by davidecaroselli** [comment]
Hi @bittlingmayer 

please note that this is supported only on the cloud-based API of ModernMT, which is a **different API** than the open-source. This is because that solution is implemented at the **cluster-level** and not per-machine where it could be basically pointless as discussed in an old post.

Thank you!
